File: scripts/analyze-leaks.py
#!/usr/bin/env python3
"""Heap leak analysis for Ambient Circuit.
Usage: python3 scripts/analyze-leaks.py [snapshot.heapsnapshot]
Default: .heap-snapshots/LEAKS.heapsnapshot
"""

import sys
import json
import os
from collections import defaultdict, deque

path = sys.argv[1] if len(sys.argv) > 1 else '.heap-snapshots/LEAKS.heapsnapshot'
size_mb = os.path.getsize(path) / 1024 / 1024
print(f'Loading {path} ({size_mb:.1f} MB)...')

with open(path, 'rb') as f:
    data = f.read()
print('Parsing JSON...')
snap = json.loads(data)
del data

nodes = snap['nodes']
edges = snap['edges']
strings = snap['strings']
meta = snap['snapshot']['meta']

# --- Field offsets ---
nf = meta['node_fields']
N = len(nf)
nType = nf.index('type')
nName = nf.index('name')
nSelfSize = nf.index('self_size')
nEdgeCount = nf.index('edge_count')
nodeTypes = meta['node_types'][0]

ef = meta['edge_fields']
E = len(ef)
eType = ef.index('type')
eName = ef.index('name_or_index')
eToNode = ef.index('to_node')
edgeTypes = meta['edge_types'][0]

nodeCount = len(nodes) // N
edgeCount = len(edges) // E
print(f'Nodes: {nodeCount:,}  Edges: {edgeCount:,}  Strings: {len(strings):,}')

def node_name(i): return strings[nodes[i * N + nName]]
def node_type(i): return nodeTypes[nodes[i * N + nType]]
def node_size(i): return nodes[i * N + nSelfSize]

# --- Build edge start index ---
print('Building edge index...')
edge_start = [0] * (nodeCount + 1)
idx = 0
for i in range(nodeCount):
    edge_start[i] = idx
    idx += nodes[i * N + nEdgeCount]
edge_start[nodeCount] = idx

# Determine owning node for edge e (binary search)
def edge_owner(e):
    lo, hi = 0, nodeCount - 1
    while lo < hi:
        mid = (lo + hi + 1) >> 1
        if edge_start[mid] <= e:
            lo = mid
        else:
            hi = mid - 1
    return lo

# --- Build retainer graph ---
print('Building retainer graph...')
retainers = defaultdict(list)  # to_node_idx -> [(from_node_idx, edge_label)]

for e in range(edgeCount):
    base = e * E
    to_idx = edges[base + eToNode] // N
    from_idx = edge_owner(e)
    et = edgeTypes[edges[base + eType]]
    en_raw = edges[base + eName]
    if et in ('element', 'hidden'):
        en = f'[{en_raw}]'
    else:
        en = strings[en_raw] if en_raw < len(strings) else f'[{en_raw}]'
    retainers[to_idx].append((from_idx, f'.{en}({et})'))

# --- Scene IDs ---
ALL_SCENES = [
    'scene-test', 'scene-structure', 'scene-rings', 'scene-instruments',
    'scene-orientation', 'scene-logic', 'scene-easing', 'scene-collisions',
    'scene-create-destroy', 'scene-rail-switch', 'scene-ctx-test',
    'scene-global-beat', 'scene-audio', 'scene-sampler-pad-deeper', 'scene-sampler-pad',
    'scene-multi-columns', 'scene-marbles', 'scene-rails', 'scene-crossing',
    'scene-reverse', 'scene-active-visible', 'scene-inactive-rails',
]
CURRENT = 'scene-sampler-pad-deeper'

import re as _re
_SCENE_RE = _re.compile(r'scene-[a-z0-9]+(?:-[a-z0-9]+)*')

def extract_scene(name):
    # First try known list
    for sid in ALL_SCENES:
        if sid != CURRENT and sid in name:
            return sid
    # Fallback: any scene-* pattern in the name that isn't current
    m = _SCENE_RE.search(name)
    if m:
        sid = m.group(0)
        if sid != CURRENT:
            return sid
    return None

# --- BFS shortest retainer path ---
def is_root(i):
    return i == 0 or not retainers[i] or node_type(i) in ('(GC roots)', 'synthetic')

def shortest_path(target, max_depth=6):
    visited = {target}
    q = deque([(target, [])])
    while q:
        idx, path = q.popleft()
        if is_root(idx) or len(path) >= max_depth:
            return path
        for (fr, label) in retainers[idx]:
            if fr in visited:
                continue
            visited.add(fr)
            step = f'{node_type(fr)}[{node_name(fr)}]{label}'
            q.append((fr, path + [step]))
    return None

# --- Find leaked objects by scene ID in name ---
print('\nSearching for leaked scene objects...')
leaked_by_scene = defaultdict(list)
for i in range(nodeCount):
    name = node_name(i)
    sid = extract_scene(name)
    if sid:
        leaked_by_scene[sid].append({'idx': i, 'name': name, 'type': node_type(i), 'size': node_size(i)})

# --- Broad type counts ---
# node_name() gives constructor name (e.g. 'BufferGeometry')
# node_type() gives V8 category ('object', 'string', etc.) — NOT the constructor
LEAKED_CONSTRUCTORS = {
    'BufferGeometry', '_BufferGeometry', 'Material', 'MeshBasicNodeMaterial', 'MeshStandardMaterial',
    'MeshBasicMaterial', 'MeshPhysicalMaterial', 'ShaderMaterial', 'Texture',
    'RenderObject',  # WebGPU renderer internal
}
broad = defaultdict(lambda: {'count': 0, 'size': 0, 'samples': []})
for i in range(nodeCount):
    constructor = node_name(i)
    if constructor not in LEAKED_CONSTRUCTORS:
        continue
    t = node_type(i)
    key = constructor
    g = broad[key]
    g['count'] += 1
    g['size'] += node_size(i)
    if len(g['samples']) < 2:
        g['samples'].append(i)
    g['_type'] = t
    g['_name'] = constructor

# Also search for BufferGeometry instances by geometry name (scene-* in .name property)
# These are found via object nodes named 'BufferGeometry' — check their children for 'name'
print('Searching for leaked BufferGeometry by geometry name...')
# Build: for each BufferGeometry node, find its .name string child
geo_by_scene = defaultdict(list)
for i in range(nodeCount):
    if node_name(i) not in ('BufferGeometry', '_BufferGeometry'):
        continue
    # scan outgoing edges for 'name' property
    es = edge_start[i]
    ee = edge_start[i + 1]
    for e in range(es, ee):
        base = e * E
        et = edgeTypes[edges[base + eType]]
        en_raw = edges[base + eName]
        en = strings[en_raw] if et not in ('element', 'hidden') and en_raw < len(strings) else None
        if en == 'name':
            to_idx = edges[base + eToNode] // N
            geo_name = node_name(to_idx)
            sid = extract_scene(geo_name)
            if sid:
                geo_by_scene[sid].append({'idx': i, 'geo_name': geo_name, 'size': node_size(i)})
            break

def fmt(n):
    if n >= 1024*1024: return f'{n/1024/1024:.1f} MB'
    if n >= 1024: return f'{n//1024} KB'
    return f'{n} B'

# ─── REPORT ──────────────────────────────────────────────────────────────────

print()
print('═' * 60)
print('LEAKED OBJECTS BY SCENE ID (in node name)')
print('═' * 60)

if not leaked_by_scene:
    print('(none found — geometry names may not embed scene IDs)')
else:
    for sid, items in sorted(leaked_by_scene.items(), key=lambda x: -len(x[1])):
        total = sum(x['size'] for x in items)
        print(f'\n── {sid}  ({len(items)} objects, {fmt(total)}) ──')
        for x in items[:5]:
            print(f'  [{x["type"]}] {x["name"]}')
            path = shortest_path(x['idx'])
            if path:
                for step in path:
                    print(f'    ← {step}')
            else:
                print('    ← (root or no path)')
        if len(items) > 5:
            print(f'  ... and {len(items)-5} more')

print()
print('═' * 60)
print('LEAKED BUFFERGEOMETRY BY .name (scene-* from other scenes)')
print('═' * 60)

if not geo_by_scene:
    print('(none found)')
else:
    for sid, items in sorted(geo_by_scene.items(), key=lambda x: -len(x[1])):
        total = sum(x['size'] for x in items)
        print(f'\n── {sid}  ({len(items)} geometries, {fmt(total)}) ──')
        for x in items[:5]:
            print(f'  BufferGeometry "{x["geo_name"]}" (node #{x["idx"]})')
            path = shortest_path(x['idx'])
            if path:
                for step in path:
                    print(f'    ← {step}')
            else:
                print('    ← (root or no path)')
        if len(items) > 5:
            print(f'  ... and {len(items)-5} more')

print()
print('═' * 60)
print('ALL GEOMETRY/MATERIAL INSTANCES (broad, by count)')
print('═' * 60)

sorted_broad = sorted(broad.values(), key=lambda g: -g['count'])
print('\nTop 20:')
for g in sorted_broad[:20]:
    print(f'  {g["count"]:5d}x  [{g["_type"]}] "{g["_name"]}"  {fmt(g["size"])}')

print()
print('═' * 60)
print('RETAINER CHAINS — top 3 geometry/material groups')
print('═' * 60)

for g in sorted_broad[:3]:
    print(f'\n[{g["_type"]}] "{g["_name"]}" — {g["count"]} instances')
    for idx in g['samples']:
        print(f'  Node #{idx}:')
        path = shortest_path(idx)
        if path:
            for step in path:
                print(f'    ← {step}')
        else:
            print('    ← (root or no path)')

print('\nDone.')

File: scripts/analyze-leaks.js
#!/usr/bin/env node
// Heap leak analysis for Ambient Circuit
// Usage: node --max-old-space-size=8192 scripts/analyze-leaks.js [snapshot.heapsnapshot]
// Default: .heap-snapshots/LEAKS.heapsnapshot

import { readFileSync } from 'fs'

const snapshotPath = process.argv[2] ?? '.heap-snapshots/LEAKS.heapsnapshot'

console.log(`Loading ${snapshotPath}...`)
const raw = readFileSync(snapshotPath)
console.log(`File size: ${(raw.length / 1024 / 1024).toFixed(1)} MB`)
console.log('Parsing JSON (may take 30-60s for large snapshots)...')

let snap
try {
	snap = JSON.parse(raw)
} catch (e) {
	console.error('JSON.parse failed:', e.message)
	console.error('Try: node --max-old-space-size=8192 scripts/analyze-leaks.js')
	process.exit(1)
}

const { nodes, edges, strings, snapshot } = snap
const meta = snapshot.meta

// --- Node field offsets ---
const nFields = meta.node_fields
const nStride = nFields.length
const nType = nFields.indexOf('type')
const nName = nFields.indexOf('name')
const nSelfSize = nFields.indexOf('self_size')
const nEdgeCount = nFields.indexOf('edge_count')

const nodeTypes = meta.node_types[0]

// --- Edge field offsets ---
const eFields = meta.edge_fields
const eStride = eFields.length
const eType = eFields.indexOf('type')
const eName = eFields.indexOf('name_or_index')
const eToNode = eFields.indexOf('to_node')

const edgeTypes = meta.edge_types[0]

const nodeCount = nodes.length / nStride
console.log(
	`Nodes: ${nodeCount.toLocaleString()}, Edges: ${(edges.length / eStride).toLocaleString()}, Strings: ${strings.length.toLocaleString()}`
)

// --- Node helpers ---
function nodeName(i) {
	return strings[nodes[i * nStride + nName]]
}
function nodeTypeName(i) {
	return nodeTypes[nodes[i * nStride + nType]]
}
function nodeSelfSize(i) {
	return nodes[i * nStride + nSelfSize]
}

// --- Build edge start index per node ---
console.log('Building edge index...')
const edgeStartArr = new Int32Array(nodeCount + 1)
{
	let eIdx = 0
	for (let i = 0; i < nodeCount; i++) {
		edgeStartArr[i] = eIdx
		eIdx += nodes[i * nStride + nEdgeCount]
	}
	edgeStartArr[nodeCount] = eIdx
}

// Map edge index → owning node index (binary search)
function edgeOwner(eIdx) {
	let lo = 0,
		hi = nodeCount - 1
	while (lo < hi) {
		const mid = (lo + hi + 1) >> 1
		if (edgeStartArr[mid] <= eIdx) lo = mid
		else hi = mid - 1
	}
	return lo
}

// --- Build retainer graph ---
console.log('Building retainer graph...')
// Use arrays-of-arrays — stored as flat typed arrays for memory efficiency
// retainerFrom[i] = list of node indices pointing to node i
// Store as Array<number[]> — acceptable for analysis script
const retainers = /** @type {{from:number,edgeName:string}[][]} */ (new Array(nodeCount))
for (let i = 0; i < nodeCount; i++) retainers[i] = []

const totalEdges = edges.length / eStride
for (let e = 0; e < totalEdges; e++) {
	const base = e * eStride
	const toNodeIdx = edges[base + eToNode] / nStride
	const fromNodeIdx = edgeOwner(e)
	const etStr = edgeTypes[edges[base + eType]]
	const enRaw = edges[base + eName]
	const enStr =
		etStr === 'element' || etStr === 'hidden' ? `[${enRaw}]` : (strings[enRaw] ?? `[${enRaw}]`)
	retainers[toNodeIdx].push({ from: fromNodeIdx, edgeName: `.${enStr}(${etStr})` })
}

// --- Scene IDs ---
const ALL_SCENE_IDS = [
	'scene-test',
	'scene-structure',
	'scene-rings',
	'scene-instruments',
	'scene-orientation',
	'scene-logic',
	'scene-easing',
	'scene-collisions',
	'scene-create-destroy',
	'scene-rail-switch',
	'scene-ctx-test',
	'scene-global-beat',
	'scene-audio',
	'scene-sampler-pad',
	'scene-sampler-pad-deeper'
]
const CURRENT_SCENE = 'scene-sampler-pad-deeper'

function extractSceneId(name) {
	for (const id of ALL_SCENE_IDS) {
		if (id !== CURRENT_SCENE && name.includes(id)) return id
	}
	return null
}

// --- BFS retainer path to root ---
function isGcRoot(nodeIdx) {
	const t = nodeTypeName(nodeIdx)
	return nodeIdx === 0 || t === '(GC roots)' || t === 'synthetic' || retainers[nodeIdx].length === 0
}

function shortestRetainerPath(targetIdx, maxDepth = 6) {
	const visited = new Set([targetIdx])
	// queue entries: {idx, path: string[]}
	const queue = [{ idx: targetIdx, path: [] }]
	while (queue.length > 0) {
		const { idx, path } = queue.shift()
		if (isGcRoot(idx) || path.length >= maxDepth) return path
		for (const { from, edgeName } of retainers[idx]) {
			if (visited.has(from)) continue
			visited.add(from)
			const label = `${nodeTypeName(from)}[${nodeName(from)}]${edgeName}`
			queue.push({ idx: from, path: [...path, label] })
		}
	}
	return null
}

// --- Find leaked objects by scene ID in name ---
console.log('\nSearching for leaked scene objects...')
const leakedByScene = new Map()

for (let i = 0; i < nodeCount; i++) {
	const name = nodeName(i)
	const sceneId = extractSceneId(name)
	if (sceneId) {
		const arr = leakedByScene.get(sceneId) ?? []
		arr.push({ nodeIdx: i, name, typeName: nodeTypeName(i), selfSize: nodeSelfSize(i) })
		leakedByScene.set(sceneId, arr)
	}
}

// --- Broad type-based counts ---
const LEAKED_TYPES = new Set([
	'BufferGeometry',
	'Material',
	'MeshBasicNodeMaterial',
	'MeshStandardMaterial',
	'MeshBasicMaterial',
	'MeshPhysicalMaterial',
	'ShaderMaterial',
	'Texture'
])

const broadGroups = new Map()
for (let i = 0; i < nodeCount; i++) {
	const type = nodeTypeName(i)
	if (!LEAKED_TYPES.has(type)) continue
	const name = nodeName(i)
	const key = `${type}::${name}`
	const g = broadGroups.get(key) ?? { type, name, count: 0, totalSize: 0, samples: [] }
	g.count++
	g.totalSize += nodeSelfSize(i)
	if (g.samples.length < 2) g.samples.push(i)
	broadGroups.set(key, g)
}

// ─── REPORT ───────────────────────────────────────────────────────────────────

function fmtBytes(n) {
	if (n >= 1024 * 1024) return `${(n / 1024 / 1024).toFixed(1)} MB`
	if (n >= 1024) return `${(n / 1024).toFixed(0)} KB`
	return `${n} B`
}

console.log()
console.log('═'.repeat(60))
console.log('LEAKED OBJECTS BY SCENE ID (in node name)')
console.log('═'.repeat(60))

if (leakedByScene.size === 0) {
	console.log('(none found — geometry names may not embed scene IDs)')
} else {
	for (const [sceneId, items] of [...leakedByScene].sort((a, b) => b[1].length - a[1].length)) {
		const totalSize = items.reduce((s, x) => s + x.selfSize, 0)
		console.log(`\n── ${sceneId}  (${items.length} objects, ${fmtBytes(totalSize)}) ──`)
		for (const { nodeIdx, name, typeName } of items.slice(0, 5)) {
			console.log(`  [${typeName}] ${name}`)
			const path = shortestRetainerPath(nodeIdx)
			if (path?.length > 0) {
				for (const step of path) console.log(`    ← ${step}`)
			} else {
				console.log(`    ← (no retainer path / root)`)
			}
		}
		if (items.length > 5) console.log(`  ... and ${items.length - 5} more`)
	}
}

console.log()
console.log('═'.repeat(60))
console.log('ALL GEOMETRY/MATERIAL INSTANCES (broad, by count)')
console.log('═'.repeat(60))

const sortedBroad = [...broadGroups.values()].sort((a, b) => b.count - a.count)
console.log(`\nTop 20:`)
for (const g of sortedBroad.slice(0, 20)) {
	console.log(
		`  ${String(g.count).padStart(5)}x  [${g.type}] "${g.name}"  ${fmtBytes(g.totalSize)}`
	)
}

console.log()
console.log('═'.repeat(60))
console.log('RETAINER CHAINS — top 3 geometry/material groups')
console.log('═'.repeat(60))

for (const g of sortedBroad.slice(0, 3)) {
	console.log(`\n[${g.type}] "${g.name}" — ${g.count} instances`)
	for (const nodeIdx of g.samples) {
		console.log(`  Node #${nodeIdx}:`)
		const path = shortestRetainerPath(nodeIdx)
		if (path?.length > 0) {
			for (const step of path) console.log(`    ← ${step}`)
		} else {
			console.log(`    ← (root or no path)`)
		}
	}
}

console.log('\nDone.')
